apiserver/params/apierror.go

```diff
@@ -52,12 +33,71 @@ func (e Error) ErrorCode() string {
 	return e.Code
 }
 
+// ErrorInfo implements the rpc.ErrorInfoProvider interface which enables
+// API error attachments to be returned as part of RPC error responses.
+func (e Error) ErrorInfo() map[string]interface{} {
+	return e.Info
+}
+
 // GoString implements fmt.GoStringer.  It means that a *Error shows its
 // contents correctly when printed with %#v.
 func (e Error) GoString() string {
 	return fmt.Sprintf("&params.Error{Message: %q, Code: %q}", e.Message, e.Code)
 }
 
+// UnmarshalInfo attempts to unmarshal the information contained in the Info
+// field of a RequestError into an AdditionalErrorInfo instance a pointer to
+// which is passed via the to argument. The method will return an error if a
+// non-pointer arg is provided.
+func (e Error) UnmarshalInfo(to interface{}) error {
+	if reflect.ValueOf(to).Kind() != reflect.Ptr {
+		return errors.New("UnmarshalInfo expects a pointer as an argument")
+	}
+
+	data, err := json.Marshal(e.Info)
+	if err != nil {
+		return errors.Annotate(err, "could not marshal error information")
+	}
+	err = json.Unmarshal(data, to)
+	if err != nil {
+		return errors.Annotate(err, "could not unmarshal error information to provided target")
+	}
+
+	return nil
+}
+
+// DischargeRequiredErrorInfo provides additional macaroon information for
+// DischargeRequired errors. Note that although these fields are compatible
+// with the same fields in httpbakery.ErrorInfo, the Juju API server does not
+// implement endpoints directly compatible with that protocol because the error
+// response format varies according to the endpoint.
+type DischargeRequiredErrorInfo struct {
+	// Macaroon may hold a macaroon that, when
+	// discharged, may allow access to the juju API.
+	// This field is associated with the ErrDischargeRequired
+	// error code.
+	Macaroon *macaroon.Macaroon `json:"macaroon,omitempty"`
+
+	// MacaroonPath holds the URL path to be associated
+	// with the macaroon. The macaroon is potentially
+	// valid for all URLs under the given path.
+	// If it is empty, the macaroon will be associated with
+	// the original URL from which the error was returned.
+	MacaroonPath string `json:"macaroon-path,omitempty"`
+}
+
+// AsMap encodes the error info as a map that can be attached to an Error.
+func (e DischargeRequiredErrorInfo) AsMap() map[string]interface{} {
+	out := make(map[string]interface{})
+	if e.Macaroon != nil {
```

Comment:
Hmm... my primary concern here is the duplication of the map keys with the serialization directives.

Given that this isn't a frequent code path, I think I'd be happier if we did the same trick as above, serialize the struct to json, then back into a map. That way we could write a generic function for it and use it for the redirect type too when it comes, but also there is a nice symetry with how we handle the conversions.

Reply:
I opted for duplication vs json marshal/unmarshal to avoid masking json-related errors which might introduce subtle errors if for example someone decides to add a field that implements `MarshalJSON`. I could change the signature to return the error but given the way it is used in tests it would be a pain to refactor.

If we don't anticipate this to be an issue I can just change the implementation as you recommend. Thoughts?

Reply:
In other places where we don't expect any error, and so don't have an error return, but internally there could be one (in theory), what we do is have something like:

`logger.Criticalf("programming error: %v", err)`

I do think that dancing through the JSON marshalling would be better here.